parquet/src/main/java/org/apache/iceberg/parquet/ParquetTypeWithPartnerVisitor.java

```diff
@@ -0,0 +1,256 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.parquet;
+
+import java.util.Deque;
+import java.util.List;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.Pair;
+import org.apache.parquet.schema.GroupType;
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.OriginalType;
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.Type;
+
+public abstract class ParquetTypeWithPartnerVisitor<P, T> {
+  private final Deque<String> fieldNames = Lists.newLinkedList();
+
+  public static <P, T> T visit(P partnerType, Type type, ParquetTypeWithPartnerVisitor<P, T> visitor) {
+    if (type instanceof MessageType) {
```

Comment:
Added.

Reply:
@JingsongLi , The `visit` is also called by `vistitList` and `visitMap` where we cannot guarantee that the inner type is not null. So I think we should allow null partner here. Does that make sense to you?

Reply:
If it's possible to have `null` for the `type`, then the `if{} else if{} else {}`  will throw NPE  ? Because they did not do any nullable check for `type`.

Reply:
OK, I got the wrong thing. Pls ignore my above comment.